### ISSUE TITLE
Juniper performance is too slow

### ISSUE THREAD

**Issue by ktbyers** [ktbyers/netmiko#2705]
See performance graphs and look into the cause of this.

**Comment by applericky** [comment]
Hey I use Juniper a lot. Let me know if you want me to do any tests.

**Comment by ktbyers** [comment]
@applericky Okay, great.

Here is the latest graph which includes Juniper. I was a bit surprised Juniper was so relatively slow.

It is a bit unclear to me whether this is due to my virtual MXs being slow or the actual Netmiko code being slow. I will send some more details on the performance test setup and how this could be replicated against real devices.

**Comment by applericky** [comment]
Hi,

Sorry I don't see the graph? Maybe its in a different spot? 

I can test both vSRX's and physical SRX's in prod.

**Comment by ktbyers** [comment]
My bad, I forgot to paste it:

https://github.com/ktbyers/netmiko/blob/develop/tests/performance/performance_report.md

I will put in some details about how I normally test this as well.

**Comment by applericky** [comment]
Cool got it. I can see the code for it too which looks nice for the graph. When the time permits send over the details on setting up.

**Comment by ktbyers** [comment]
You can comment out the large ACL test and focus on the connect, show. The small config change should be benign. You can set the test device in the the test_devices.yml file.

You can comment out these two lines two disable the configuration changes.

https://github.com/ktbyers/netmiko/blob/develop/tests/performance/test_netmiko.py#L191-L192

The main way I run the test is by running this script here:

https://github.com/ktbyers/netmiko/blob/develop/tests/performance/test_performance.sh

You can see what things that script executes (i.e. that test_performance.sh is just a simple wrapper script).

**Comment by applericky** [comment]
@ktbyers I have the repo downloaded and I have done most of the setup. Had to change some things to fit my environment from yours. I just now need to work on the gen_graph.py.It's failing the tests on the split(":"). I'm getting a nonetype error. I'll keep you posted on progress.

**Comment by applericky** [comment]
Attempted again.  Having some issues with def convert_time(time_str):
Here is what I'm seeing.


```

--------------------------------------------------------------------------------
Device name: juniper_vmx
------------
<function connect at 0x10a16fb80>: Elapsed time: 0:00:06.358340
<function send_command_simple at 0x10a16fca0>: Elapsed time: 0:00:07.185749
<function cleanup at 0x10a171430>: Elapsed time: 0:00:17.445482
--------------------------------------------------------------------------------

```

```
File "netmiko-develop/tests/performance/gen_graph.py", line 16, in convert_time
    results = time_str.split(":")
AttributeError: 'NoneType' object has no attribute 'split'
```

Looks like I'm right at the last part. As mentioned before I have made some changes to a few things to just get it started but this part looks like its requiring some changes? I don't want to mess with the code too much so the reading I give you are accurate.

**Comment by ktbyers** [comment]
We don't really need the graphs for this issue. The main problem/item here is figuring out why it takes 6.35 seconds to just connect or 7.2 seconds to execute a simple show command. In other words, we should be able to just look and use the numbers we get back from the test.

We probably should look into the session_preparation method in the Juniper class and see if we can figure out why it is taking so long (anyways that is where I would look).

**Comment by applericky** [comment]
Ahh ok I thought we needed the graphs. I tested it against a vSRX. I'll test out some different models of SRX.

**Comment by ktbyers** [comment]
Yeah, it is all good. The main issue is why is it taking so long (as 6.4 seconds just to login seems pretty slow).

**Comment by applericky** [comment]
Copy. I'll take a look too.

**Comment by ktbyers** [comment]
You can use this decorator here to show the time that various things take to execute when called:

https://github.com/ktbyers/netmiko/blob/develop/netmiko/utilities.py#L600

The m_exec_time is for methods. There is a similar f_exec_time for straight functions.

You would just import it and then put it before the method. For example:

```
@m_exec_time
def my_method(self):
```

I would probably look at the things that happen in the session_preparation method i.e. here:

https://github.com/ktbyers/netmiko/blob/develop/netmiko/juniper/juniper.py#L18

Basically, see if we can figure out what is taking the bulk of that 6.4 seconds in connect time.

**Comment by applericky** [comment]
To confirm, 

I added the following to https://github.com/ktbyers/netmiko/blob/develop/netmiko/juniper/juniper.py#L18

```
 @m_exec_time
    def session_preparation(self) -> None:
```

I also verified that it was already imported.

`from netmiko import utilities`


- Tested the test script again and nothing populated.
- Tested by removing the version of Netmiko on my system and installed the netmiko-develop version. 

I ran an existing script that I had to see if there would be anything there but nothing came up. I think either the import is not correct or the decorator isn't being called. 

Do you have an existing example somewhere to test against?

**Comment by ktbyers** [comment]
I would try doing the import like this:

```python
from netmiko.utilities import m_exec_time
```

**Comment by ktbyers** [comment]
@applericky Looking into this some more and the issue ended up being due to problems on `disconnect()`.

I made a PR that improves this quite a bit here:

https://github.com/ktbyers/netmiko/pull/2753

**Comment by applericky** [comment]
Can't wait to try it!
